### PR TITLE
qemu: Enable more via-mpy and via-mpy with native emitter tests

### PR DESCRIPTION
**Pull request by dpgeorge**
### Summary

The qemu port has recently undergone a lot of improvement, including adding of RISC-V support and improved test running to use the standard bare-metal test approach.

One of the reasons to improve the test runner was to be able to enable more native emitter tests on the qemu port, by leveraging the `--via-mpy` and `--emit native` arguments to `run-tests.py`.

That can now be done, and is now done in this PR.

### Testing

Tested locally with `make test_full`.  Updated CI to use this new `test_full` make target.

### Trade-offs and Alternatives

This will make CI a bit slower but not by much, and the benefit is well worth it (better test coverage of the ARM and RISC-V native emitters).

### CODE REVIEW

**Comment by dpgeorge** [comment]
@agatti here's the final piece of the puzzle to get more native tests on the qemu port.

**Comment by dpgeorge** [comment]
The CI shows the following failures:
- Cortex-A9 has issue with some native tests, perhaps the ARM (non-Thumb) emitter has some bugs?
- On RISC-V, the `extmod/deflate_decompress.py` test fails in native mode.

**Comment by agatti** [comment]
> On RISC-V, the `extmod/deflate_decompress.py` test fails in native mode.

It can also fail using the VM.  See [`afb0102` (#15714)](https://github.com/micropython/micropython/pull/15714/commits/afb0102524a97722a25f2bcb35c98ae990231655).

**Comment by agatti** [comment]
@dpgeorge can this handle natmods or is something like #15851 still needed for importing MPY files from the filesystem?

**Comment by dpgeorge** [comment]
> can this handle natmods or is something like #15851 still needed for importing MPY files from the filesystem?

It could handle the natmod examples with a few more tweaks, eg add a RAM VFS and copy the .mpy files into that to be executed.

Or #15851 is another way to do it.

**Comment by agatti** [comment]
I see, thanks!  If you want I can take out the GC heap size commit out of the inline assembler PR and submit it as its own thing for you to merge and fix the RV32 failure.

**Comment by agatti** [comment]
This is weird, I see the same error I got when trying to run the RV64 tests but with the SABRELITE target...

```bash
[mpy-dev-arm /micropython/ports/qemu]$ make BOARD=SABRELITE DEBUG=1 test RUN_TESTS_ARGS="--via-mpy --emit native" V=1

python3 ../../py/makeversionhdr.py build-SABRELITE/genhdr/mpversion.h
test -e "../../lib/micropython-lib/README.md" || (echo -e "\033[1;31mError: micropython-lib submodule is not initialized.\033[0m Run 'make submodules'"; false)
python3 ../../tools/makemanifest.py -o build-SABRELITE/frozen_content.c -v "PORT_DIR=/micropython/ports/qemu" -v "MPY_DIR=../.." -v "BOARD_DIR=" -v "MPY_LIB_DIR=../../lib/micropython-lib" -b "build-SABRELITE" -f"-march=armv6" --mpy-tool-flags="" "freeze('test-frzmpy')"
cd ../../tests && ./run-tests.py --target qemu --device execpty:"qemu-system-arm -m 128M -machine sabrelite -nographic -monitor null -semihosting  -serial pty -kernel ../ports/qemu/build-SABRELITE/firmware.elf" --via-mpy --emit native 
Traceback (most recent call last):
  File "/micropython/tests/./run-tests.py", line 1149, in <module>
    main()
  File "/micropython/tests/./run-tests.py", line 1139, in main
    res = run_tests(pyb, tests, args, args.result_dir, args.jobs)
  File "/micropython/tests/./run-tests.py", line 859, in run_tests
    run_one_test(test)
  File "/micropython/tests/./run-tests.py", line 824, in run_one_test
    output_mupy = run_micropython(pyb, args, test_file, test_file_abspath)
  File "/micropython/tests/./run-tests.py", line 312, in run_micropython
    had_crash, output_mupy = pyb.run_script_on_remote_target(
  File "/micropython/tests/./run-tests.py", line 170, in run_script_on_remote_target
    output_mupy = pyb.exec_(script)
  File "/micropython/tools/pyboard.py", line 496, in exec_
    ret, ret_err = self.exec_raw(command, data_consumer=data_consumer)
  File "/micropython/tools/pyboard.py", line 482, in exec_raw
    return self.follow(timeout, data_consumer)
  File "/micropython/tools/pyboard.py", line 390, in follow
    data = self.read_until(1, b"\x04", timeout=timeout, data_consumer=data_consumer)
  File "/micropython/tools/pyboard.py", line 335, in read_until
    elif self.serial.inWaiting() > 0:
  File "/micropython/tools/pyboard.py", line 263, in inWaiting
    return self.serial.inWaiting()
  File "/usr/lib/python3/dist-packages/serial/serialutil.py", line 594, in inWaiting
    return self.in_waiting
  File "/usr/lib/python3/dist-packages/serial/serialposix.py", line 549, in in_waiting
    s = fcntl.ioctl(self.fd, TIOCINQ, TIOCM_zero_str)
OSError: [Errno 5] Input/output error
make: *** [Makefile:170: test] Error 1
```

I'm running this from inside LXC containers but they used to work for MicroPython development for the past two years.

**Comment by agatti** [comment]
This unblocks testing on SABRELITE (TL;DR: it generated LDRH Rd, [Rn, #imm]!, so the second LDRH from the same origin would load from the wrong base).  Now tests pass until it gets to `fun_largestate.py`, where it crashes.  I'll look at it later and will send a PR once the test suite passes...

```diff
diff --git a/py/asmarm.c b/py/asmarm.c
index 600649070..634f2db4f 100644
--- a/py/asmarm.c
+++ b/py/asmarm.c
@@ -328,7 +328,7 @@ void asm_arm_ldrh_reg_reg(asm_arm_t *as, uint rd, uint rn) {

 void asm_arm_ldrh_reg_reg_offset(asm_arm_t *as, uint rd, uint rn, uint byte_offset) {
     // ldrh rd, [rn, #off]
-    emit_al(as, 0x1f000b0 | (rn << 16) | (rd << 12) | ((byte_offset & 0xf0) << 4) | (byte_offset & 0xf));
+    emit_al(as, 0x1d000b0 | (rn << 16) | (rd << 12) | ((byte_offset & 0xf0) << 4) | (byte_offset & 0xf));
 }

 void asm_arm_ldrb_reg_reg(asm_arm_t *as, uint rd, uint rn) {
```

**Comment by dpgeorge** [comment]
> This unblocks testing on SABRELITE

Thanks!  I applied your patch.

**Comment by agatti** [comment]
> Now tests pass until it gets to `fun_largestate.py`, where it crashes. 

`basics/fun_largestate.py` requires a stack larger than 256 bytes, which is not handled by `asm_arm_entry` and `asm_arm_exit`.  They defer stack adjustment to a simple register sub/add opcode with an embedded immediate that cannot represent the required adjustment.

**Comment by dpgeorge** [comment]
> `basics/fun_largestate.py` requires a stack larger than 256 bytes, which is not handled by `asm_arm_entry` and `asm_arm_exit`. They defer stack adjustment to a simple register sub/add opcode with an embedded immediate that cannot represent the required adjustment.

Good debugging!

This is exactly why extensive tests are good, they catch these edge cases.

**Comment by agatti** [comment]
And this fixes `basics/fun_largestate.py`.

```diff
diff --git a/py/asmarm.c b/py/asmarm.c
index 634f2db4f..5da719ff3 100644
--- a/py/asmarm.c
+++ b/py/asmarm.c
@@ -168,15 +168,24 @@ void asm_arm_entry(asm_arm_t *as, int num_locals) {
 
     emit_al(as, asm_arm_op_push(as->push_reglist | 1 << ASM_ARM_REG_LR));
     if (as->stack_adjust > 0) {
-        emit_al(as, asm_arm_op_sub_imm(ASM_ARM_REG_SP, ASM_ARM_REG_SP, as->stack_adjust));
+       if (as->stack_adjust < 0x100) {
+            emit_al(as, asm_arm_op_sub_imm(ASM_ARM_REG_SP, ASM_ARM_REG_SP, as->stack_adjust));
+       } else {
+           asm_arm_mov_reg_i32_optimised(as, ASM_ARM_REG_R4, as->stack_adjust);
+           emit_al(as, asm_arm_op_sub_reg(ASM_ARM_REG_SP, ASM_ARM_REG_SP, ASM_ARM_REG_R4));
+       }
     }
 }
 
 void asm_arm_exit(asm_arm_t *as) {
     if (as->stack_adjust > 0) {
-        emit_al(as, asm_arm_op_add_imm(ASM_ARM_REG_SP, ASM_ARM_REG_SP, as->stack_adjust));
+       if (as->stack_adjust < 0x100) {
+            emit_al(as, asm_arm_op_add_imm(ASM_ARM_REG_SP, ASM_ARM_REG_SP, as->stack_adjust));
+       } else {
+           asm_arm_mov_reg_i32_optimised(as, ASM_ARM_REG_R4, as->stack_adjust);
+           emit_al(as, asm_arm_op_add_reg(ASM_ARM_REG_SP, ASM_ARM_REG_SP, ASM_ARM_REG_R4));
+       }
     }

     emit_al(as, asm_arm_op_pop(as->push_reglist | (1 << ASM_ARM_REG_PC)));
 }
```

**Comment by dpgeorge** [comment]
I've applied the above patch, thank you.

**Comment by dpgeorge** [comment]
The CI here is finally green!

I had to skip a couple of tests for the SABRELITE board because they fail with the native emitter (it's using armv6 as the architecture which isn't quite right but is the best match at this stage, and some big tests just don't work with that emitter for some reason).

@agatti what do you think about this PR now, do you think it's good to merge?

**Comment by agatti** [comment]
I'd say it looks good to me!  There's just one minor thing: I'm having second thoughts on my choice of trashing R4 for larger stack frames' allocation/deallocation.  Using R8 for that is probably a safer choice providing equivalent functionality, since as far as the compiler framework is concerned that register doesn't exist (and R8 is trashed in `asm_arm_strh_reg_reg_reg` anyway).

**Comment by dpgeorge** [comment]
> I'm having second thoughts on my choice of trashing R4 for larger stack frames' allocation/deallocation. Using R8 for that is probably a safer choice going forward, providing equivalent functionality

OK, now changed to R8.
